### DISCUSSION TITLE
NodeGraphProcessor based on deprecated APIs?

### DISCUSSION THREAD

**Discussion by leissler** [alelievr/NodeGraphProcessor#144] [General]
Hi,
since you mentioned in one of the issues that the GraphView team of Unity has been all moved over to a new team building a new API (GTF), does that mean that this whole project will soon be based on deprecated APIs, which will disappear sometime in the future? 
What are your plans for maintaining the NGP? Honestly, it is the best approach for a Unity-based node framework and Unity should just have adapted this. They even have an employee who created this. All sounds very weird.

- Martin

**Comment by alelievr** [comment]
Hello,

> does that mean that this whole project will soon be based on deprecated APIs, which will disappear sometime in the future?

Yes, that's correct. When GTF will be ready the GraphView API will be deprecated and removed but I expect it to be slow because many users based their projects on the GraphView API + there will be the internal migration of all the graph tools in Unity that will take a lot of time too (VFX Graph, ShaderGraph).

> What are your plans for maintaining the NGP?

When GTF will be released and stable, I plan to migrate all the GraphView code I have in this project to the new GTF code. Though I don't know how much work / refactor it is. Ideally, I'd like to keep all the features I have right now and benefit from all the nice things GTF has.
That being said, it will probably be for NGP 2.0 or even 3.0 depending on how much time it takes for GTF to be stable.

> Honestly, it is the best approach for a Unity-based node framework and Unity should just have adapted this

Thank you :D I tried to keep the code open and readable so it gives maximum flexibility to the end-user and it seems to be paying off!

> They even have an employee who created this. All sounds very weird

I started creating this Graph project before joining Unity so it was logical for me to continue maintaining it and adding more features.  Also, at Unity, I'm a Graphics Developer in HDRP so my work is completely unrelated to any internal graph-related work :)

Finally, I'd add that a project like this one couldn't exist inside the Unity ecosystem because, in term of code structure, everything is public, which is literally the opposite of what an official Unity package would be (internal by default and then public when the API is stable, documented and tested). I continued this way simply because it gives a lot more power to the developers but I know that it also can cause more confusion or mistakes depending on the use case / context.

**Reply by leissler** [reply]
As a Unity insider, would you have any information on when GTF will hit? The [thread at Unity forums](https://forum.unity.com/threads/graph-tool-foundation.1057667/page-2) seems to be abandoned by Unity. I hope the layoffs were not in the GTF territory.

**Reply by alelievr** [reply]
Hard to know as of right now, there is a plan for 2023 but not sure it will happen. In any case, the GTF development is not stopped and more teams are working on converting existing graph-based tools to GTF

**Reply by leissler** [reply]
Hey, thanks for this estimation! How much sense does it make from your point of view to start diving into the GTF API? Do you think it will change significantly (and the concepts and methods behind it)? Or do you think the changes to the final API will be rather marginal?

**Reply by alelievr** [reply]
At this point, I think it's safe to say that the API will not change too much. You can start diving into GTF now but you may struggle to find documentation and example code.
